File: java/src/main/java/br/com/emendes/problems/SqrtX.java
package br.com.emendes.problems;

/**
 * Dado um inteiro não negativo x, retorne a raiz quadrada de x arredondada para o número inteiro mais próximo.
 * O inteiro retornado deve ser não negativo também.<br><br>
 *
 * Você não deve usar nenhuma função ou operador imbutido.<br><br>
 *
 * Restrições:<br>
 * 0 <= x <= 2³¹ - 1
 */
public class SqrtX {

  public int mySqrt(int x) {
    int min = 0;
    int max = Math.min(x, 46_340);

    int answer = 0;

    while (min <= max) {
      int mid = (max - min)/2 + min;
      int midPower = mid * mid;

      if (midPower < x) {
        answer = mid;
        min = mid + 1;
      } else if (midPower == x) {
        return mid;
      } else {
        max = mid - 1;
      }
    }

    return answer;
  }

}

File: javascript/problems/SqrtX.js
/**
 * Dado um inteiro não negativo, retorne a raiz quadrada de x arredondado para o número inteiro mais próximo.
 * O inteiro retornado também deve ser não negativo.
 *
 * Você não deve usar nenhuma função ou operador expoente integrado, como:
 * pow(x, 0.5) em c++ ou x ** 0.5 in python.
 *
 * Restrições:
 *
 * 0 <= x <= 2³¹ - 1
 *
 * @param {number} x
 * @return {number}
 */
const mySqrt = function (x) {
  let left = 0;
  let right = 46341;
  while (true) {
    const mid = Math.floor((right + left) / 2);
    if (mid * mid < x) {
      if ((mid + 1) * (mid + 1) > x) return mid;
      left = mid + 1;
    } else if (mid * mid > x) {
      if ((mid - 1) * (mid - 1) < x) return mid - 1;
      right = mid - 1;
    } else return mid;
  }
};
